### ISSUE TITLE
[ITensors] [BUG] QN correlation matrix not working since #1178

### ISSUE THREAD

**Issue by corbett5** [ITensor/ITensors.jl#1209]
**Description of bug**

Repeat of #1179

When computing the correlation matrix of two operators which when acted together on the same site give zero with a quantum number conserving MPS I get an error.

**Minimal code demonstrating the bug or unexpected behavior**

<details><summary>Minimal runnable code</summary><p>

```julia
using ITensors
opP = Int32[1 0 0 0; 0 0 0 0; 0 0 0 0; 0 0 0 0]
opQ = Int32[0 0 0 0; 0 1 0 0; 0 0 0 0; 0 0 0 0]
sites = siteinds("Electron", 4; conserve_qns=true)
psi = MPS(sites, ["Up" for n in 1:4])
correlation_matrix(psi, opP, opQ)
```
</p></details>

**Expected output or behavior**

The correlation matrix is computed.

**Actual output or behavior**

Bounds error

<details><summary>Output of minimal runnable code</summary><p>

```
ERROR: BoundsError: attempt to access 0-element Vector{Block{2}} at index [1]
Stacktrace:
  [1] getindex
    @ ./essentials.jl:13 [inlined]
  [2] first
    @ ./abstractarray.jl:445 [inlined]
  [3] checkflux(T::NDTensors.BlockSparseTensor{Int32, 2, Tuple{Index{Vector{Pair{QN, Int64}}}, Index{Vector{Pair{QN, Int64}}}}, NDTensors.BlockSparse{Int32, Vector{Int32}, 2}})
    @ ITensors ~/.julia/packages/ITensors/yZbTa/src/qn/flux.jl:50
  [4] ITensor(::NDTensors.AllowAlias, elt::Type{Int32}, A::Matrix{Int32}, inds::Vector{Index{Vector{Pair{QN, Int64}}}}; tol::Float64, checkflux::Bool)
    @ ITensors ~/.julia/packages/ITensors/yZbTa/src/qn/qnitensor.jl:316
  [5] ITensor(::NDTensors.AllowAlias, elt::Type{Int32}, A::Matrix{Int32}, inds::Vector{Index{Vector{Pair{QN, Int64}}}})
    @ ITensors ~/.julia/packages/ITensors/yZbTa/src/qn/qnitensor.jl:297
  [6] ITensor(::NDTensors.AllowAlias, ::Matrix{Int32}, ::Vector{Index{Vector{Pair{QN, Int64}}}}, ::Vararg{Vector{Index{Vector{Pair{QN, Int64}}}}}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ ITensors ~/.julia/packages/ITensors/yZbTa/src/itensor.jl:402
  [7] ITensor(::NDTensors.AllowAlias, ::Matrix{Int32}, ::Vector{Index{Vector{Pair{QN, Int64}}}}, ::Vararg{Vector{Index{Vector{Pair{QN, Int64}}}}})
    @ ITensors ~/.julia/packages/ITensors/yZbTa/src/itensor.jl:399
  [8] itensor(::Matrix{Int32}, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ ITensors ~/.julia/packages/ITensors/yZbTa/src/itensor.jl:134
  [9] itensor
    @ ~/.julia/packages/ITensors/yZbTa/src/itensor.jl:134 [inlined]
 [10] op(X::Matrix{Int32}, s::Index{Vector{Pair{QN, Int64}}})
    @ ITensors ~/.julia/packages/ITensors/yZbTa/src/physics/sitetype.jl:435
 [11] #op#1092
    @ ~/.julia/packages/ITensors/yZbTa/src/physics/sitetype.jl:464 [inlined]
 [12] op
    @ ~/.julia/packages/ITensors/yZbTa/src/physics/sitetype.jl:463 [inlined]
 [13] #op#1095
    @ ~/.julia/packages/ITensors/yZbTa/src/physics/sitetype.jl:468 [inlined]
 [14] op(opname::Matrix{Int32}, s::Vector{Index{Vector{Pair{QN, Int64}}}}, ns::Int64)
    @ ITensors ~/.julia/packages/ITensors/yZbTa/src/physics/sitetype.jl:467
 [15] correlation_matrix(psi::MPS, _Op1::Matrix{Int32}, _Op2::Matrix{Int32}; sites::UnitRange{Int64}, site_range::Nothing, ishermitian::Nothing)
    @ ITensors ~/.julia/packages/ITensors/yZbTa/src/mps/mps.jl:796
 [16] correlation_matrix(psi::MPS, _Op1::Matrix{Int32}, _Op2::Matrix{Int32})
    @ ITensors ~/.julia/packages/ITensors/yZbTa/src/mps/mps.jl:714
 [17] top-level scope
    @ REPL[12]:1
```

</p></details>

**Version information**

 - Output from `versioninfo()`:
```julia
julia> versioninfo()
Julia Version 1.9.2
Commit e4ee485e909 (2023-07-05 09:39 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 24 × 13th Gen Intel(R) Core(TM) i7-13700KF
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, goldmont)
  Threads: 1 on 24 virtual cores
Environment:
  JULIA_REVISE_POLL = 1
```
 - Output from `using Pkg; Pkg.status("ITensors")`:
```julia
julia> using Pkg; Pkg.status("ITensors")
Status `~/.julia/environments/v1.9/Project.toml`
  [9136182c] ITensors v0.3.43
```

Note that this worked as expected in v0.3.41 before #1178 added `checkflux` in the QN tensor construction.

**Comment by emstoudenmire** [comment]
Thanks Ben. I'm seeing this bug too. It looks like in the ITensor constructor around line 297 of qn/qnitensor.jl, that the BlockSparseTensor that gets made inside of there ends up not having any blocks for these inputs, though it seems it should. This causes the `checkflux(T)` call a few lines later to fail. (Just mentioning this to record what I've found so far to discuss with Karl and Matt.)

**Comment by emstoudenmire** [comment]
I just pushed a fix in PR #1210.

**Comment by corbett5** [comment]
Excellent, thank you!
